File: exJava/go/Q25.go
package main

import "fmt"

func main() {
	iter := newIterator()
	for i := 0; i < 10; i++ {
		iter.add(i)
	}
	iter.remove()
	iter.add(30)
	iter = iter.iterator()
	for iter.hasNext() {
		if i, ok := iter.next().(int); ok {
			fmt.Printf("%d\n", i)
		}
	}
}

type iterator interface {
	hasNext() bool
	next() interface{}
	remove()
	iterator() iterator
	add(interface{})
}

func newIterator() iterator {
	iter := new(myIterator)
	iter.arr = make([]interface{}, 0)
	iter.index = -1
	return iter
}

type myIterator struct {
	arr   []interface{}
	index int
}

func (iter *myIterator) add(value interface{}) {
	iter.index++
	iter.arr = append(iter.arr, value)
}
func (iter *myIterator) hasNext() bool {
	return iter.index < len(iter.arr)-1
}
func (iter *myIterator) next() interface{} {
	iter.index++
	return iter.arr[iter.index]
}
func (iter *myIterator) remove() {
	if len(iter.arr) == 0 { // try iter.arr == nil
		return
	}
	iter.arr = iter.arr[:len(iter.arr)-1]
}
func (old *myIterator) iterator() iterator {
	iter := new(myIterator)
	iter.arr = old.arr
	iter.index = -1
	return iter
}

File: shibaura-institute-technology/exJava/java/Q25.java
import java.util.*;

public class Q25 {
  public static void main(String args[]) {
    new Q25Main();
  }
}

class Q25Main {

  public Q25Main() {

    MyIterator iter = new MyIterator();

    for(int i = 0; i < 10; i++) {
      iter.add(i);
    }

    iter.remove();
    iter.add(30);

    iter = iter.iterator();
    while(iter.hasNext()) {
      System.out.println(iter.next());
    }

  }
}

// as linked list
class MyIterator implements Iterator {

  private LinkedList ll;
  private Node node;

  MyIterator() {
    ll = new LinkedList();
    node = ll.getHead();
  }
  MyIterator(MyIterator oll) {
    ll = oll.ll;
    node = ll.getHead();
  }

  public boolean hasNext() {
    return node != null && node.next != null;
  }

  public Object next() {
    if(node != null) {
      node = node.next;
      return node.value;
    }
    return null;
  }

  public void add(Object value) {
    ll.add(value);
  }

  public void remove() {
    ll.remove();
  }

  MyIterator iterator() {
    return new MyIterator(this);
  }

}



class LinkedList {

  private Node head; 

  LinkedList() {
    head = new Node(); // always empty
  }

  public void add(Object value) {
    Node node = head;
    while(node.next != null) {
      node = node.next;
    }
    node.next = new Node(value);
  }

  public void remove() {
    Node node = head;
    while(node.next != null && node.next.next != null) {
      node = node.next;
    }
    node.next = null;
  }

  public Node getHead() {
    return head;
  }

}

class Node {
  Node next; 
  Object value;

  Node(){
  }
  Node(Object value) {
    this.value = value;
  }

  public String toString() {
    return value.toString();
  }
}
